### ISSUE TITLE
Maintain path relative to root folder

### ISSUE THREAD

**Issue by thenewguy** [actions/upload-artifact#174]
It would be helpful if my artifact would not squash the directory structure.  For example,

```
- name: Archive production artifacts
        uses: actions/upload-artifact@v2
        with:
          name: assets
          retention-days: 5
          path: ./root/relative/path/to/my/file*
```

I end up with a zipfile containing `file.js` and `file.css`.  What I want is a zipfile containing `root/relative/path/to/my/file.js` and `root/relative/path/to/my/file.css`

**Comment by rcowsill** [comment]
As a workaround you could make an empty file in `root/` and include it in the archive. That should force the action to produce this directory structure:
```
empty.txt
relative
└ path
  └ to
    └ my
       file.css
       file.js
```

**Comment by konradpabjan** [comment]
This behavior is documented: https://github.com/actions/upload-artifact#upload-using-multiple-paths-and-exclusions

The reason why only `file.js` and `file.css` are included without a root directory is because you're using a wildcard pattern and it gets cutoff there.

If you do something like `path: ./*/relative/path/to/my/file*` then it should include the full path.

**Comment by aminya** [comment]
There should be an option to maintain the paths. The reason is that if you use `download-artifact` later, you have to manually move the files from the `./artifact` folder.

**Comment by aramfe** [comment]
Yeah. It's quite frustrating that this completely flattens out the path, which is essentially bad design in the context of workflows that use artifacts between different jobs. GitLab for example does it as one would expect by completely maintaining the absolute path, without any flattening. 

KISS should have been used as principle here instead of some magic shrinking of the path. At least add the option to do so.

**Comment by tulsishell** [comment]
I found a way around this - include one other file in your  artifact which doesn't have the same base path as all the other files. 
It should be in a completely different folder from the root itself. 
This way the action doesn't find any least common ancestor of all the search paths and gives you complete directory structure. :)

**Comment by mgebundy** [comment]
Thanks for the workaround @rcowsill & @tulsishell 

I would love to see a simple option for maintaining paths, it's causing similar issues on our end and I'm not even interested in using wildcards.

**Comment by asos-tomp** [comment]
> I found a way around this - include one other file in your artifact which doesn't have the same base path as all the other files. It should be in a completely different folder from the root itself. This way the action doesn't find any least common ancestor of all the search paths and gives you complete directory structure. :)

Frustratingly, an approach that can't be used with parallel jobs I assume, since unless each choses a unique "one other file", the mentioned 503 errors will kick in.

**Comment by retorquere** [comment]
> Frustratingly, an approach that can't be used with parallel jobs I assume, since unless each choses a unique "one other file", the mentioned 503 errors will kick in.

I'm running this to keep the dir structure

```
mkdir artifacts
mkdir .artifacts
mktemp .artifacts/artifacts.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```

and then specify both for upload. Which is... not great, but works.

**Comment by Naman-Kumar-Sinha** [comment]
> > Frustratingly, an approach that can't be used with parallel jobs I assume, since unless each choses a unique "one other file", the mentioned 503 errors will kick in.
> 
> I'm running this to keep the dir structure
> 
> ```
> mkdir artifacts
> mkdir .artifacts
> mktemp .artifacts/artifacts.XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
> ```
> 
> and then specify both for upload. Which is... not great, but works.

Thanks for the tip @retorquere! This worked well. I wonder if it can be an opt-in functionality in this action.

**Comment by ken-cdit** [comment]
Can't believe there is still no explicit option for this.

**Comment by dlarocque** [comment]
+1

**Comment by Perksey** [comment]
Definitely a +1 here, my scenario is having a matrix job with specific outputs (with a specific directory structure), where all outputs are then aggregated in a later job in the workflow (again expecting the directory structure to be maintained)

**Comment by jsirois** [comment]
Thanks @retorquere. I adapted your approach by just inclduing `.gitignore`. Its handily available in almost every CI setup and, in the end, it is not included in the zip even though the multi-path common ancestor trick works and preserve all other paths directory structure.
Concretely, this worked for me:
```yaml
  - uses: actions/upload-artifact@v4
    with:
      name: test-results-unit
      path: |
        dist/test-results/
        .gitignore
```

The result was a zip that preserved dir structure:
```
:; unzip -l test-results-unit.zip 
Archive:  test-results-unit.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
  2570919  2025-02-10 20:33   dist/test-results/unit.xml
---------                     -------
  2570919                     1 file
```

**Comment by irios-phorus** [comment]
4 years later, and we still have to rely on workarounds

**Comment by Rapti** [comment]
You can simply mention a dummy file in the desired archive root. If the file doesn't exist, it won't pollute your archive, but it will still be accounted for when the action tries to flatten the directory structure.

    path: |
      dummy-file-to-keep-directory-structure.txt
      dist/*

> Edit: this workaround also means I can't use `if-no-files-found: error` at all, since there will ALWAYS be an dummy file included in the upload.

Just don't actually create the dummy file and `if-no-files-found` will work as well.

**Comment by istankovic** [comment]
It's hard to believe there's still no option for this and people need to work around the problem.
